### PR TITLE
Update to use dune and prep for publishing new version to OPAM

### PR DESCRIPTION
**Pull request by reubenrowe**
@zapashcanon had already done the port to dune a few years ago. I have updated the dune-project file and bumped the version number. This should be ready to publish to OPAM now.

### CODE REVIEW

File: lib/dune [hunk 1]
```diff
@@ -0,0 +1,3 @@
+(library
+  (name hashset)
```
Comment (redianthus) [inline lib/dune]: This should be `public_name` otherwise it won't be installed.

**Comment by redianthus** [comment]
Hi @reubenrowe, long time no see. :)

I forgot I did this and never opened a PR, thanks! I'll notify Jean-Christophe about your PR in case he doesn't see it.

File: lib/dune (hunk 1 above)
Comment (reubenrowe) [reply lib/dune]: Oh yes, thanks. Done now.

Comment (redianthus) [reply lib/dune]: You can even remove the `(name hashet)` now that there's a `public_name`. :)

Comment (reubenrowe) [reply lib/dune]: Oh, OK. That is not really apparent from the documentation: https://dune.readthedocs.io/en/stable/dune-files.html#library

Comment (redianthus) [reply lib/dune]: Yes, that's because you can have both and they can be different - but that's usually a terrible idea IMO.

**Comment by ramseyharrison** [comment]
Hello, did the dune edition ever get published to opam?

**Comment by redianthus** [comment]
I don't think so. I'll ping @backtracking.

**Pull request merged**